src/core/egg-herbie.rkt

```diff
@@ -106,37 +106,29 @@
       [(list op ids ...) (egraph_add_node ptr (~s op) (list->u32vec ids))]
       [(? (disjoin symbol? number?) x) (egraph_add_node ptr (~s x) 0-vec)]))
 
-  (define mappings (build-vector (batch-length batch) values))
-  (define (remap x)
-    (vector-ref mappings x))
-
-  ; Inserting nodes bottom-up
-  (for ([node (in-batch batch)]
-        [n (in-naturals)])
-    (define idx
-      (match node
-        [(literal v _) (insert-node! v)]
-        [(? number?) (insert-node! node)]
-        [(? symbol?) (insert-node! (var->egg-var node ctx))]
-        [(hole prec spec) (remap spec)] ; "hole" terms currently disappear
-        [(approx spec impl) (insert-node! (list '$approx (remap spec) (remap impl)))]
-        [(list op (app remap args) ...) (insert-node! (cons op args))]))
-    (vector-set! mappings n idx))
-  (for ([root (in-vector (batch-roots batch))])
-    (egraph_add_root ptr (remap root)))
-
-  (for ([node (in-batch batch)]
-        #:when (approx? node))
-    (match-define (approx spec impl) node)
-    (hash-ref! id->spec
-               (remap spec)
-               (lambda ()
-                 (define spec* (normalize-spec (batch-pull batch spec)))
-                 (define type (representation-type (repr-of-node batch impl ctx)))
-                 (cons spec* type))))
-
-  (for/list ([root (in-vector (batch-roots batch))])
-    (remap root)))
+  (define add-to-egraph
+    (batch-map batch
+               (λ (get-remapping node)
+                 (match node
+                   [(literal v _) (insert-node! v)]
+                   [(? number?) (insert-node! node)]
+                   [(? symbol?) (insert-node! (var->egg-var node ctx))]
+                   [(hole prec spec) (get-remapping spec)] ; "hole" terms currently disappear
+                   [(approx spec impl)
+                    (hash-ref! id->spec ; Save original (spec, type) for extraction
+                               (get-remapping spec)
+                               (lambda ()
+                                 (define spec* (normalize-spec (batch-pull (batchref batch spec))))
+                                 (define type
+                                   (representation-type (repr-of-batchref (batchref batch impl) ctx)))
+                                 (cons spec* type)))
+                    (insert-node! (list '$approx (get-remapping spec) (get-remapping impl)))]
+                   [(list op (app get-remapping args) ...) (insert-node! (cons op args))]))))
+
+  (for/list ([brf brfs])
```

Comment:
```suggestion
  (for/list ([brf (in-list brfs)])
```
Adding the "in-foo" annotations makes the for loop much faster for stupid Racket reasons, and this code here is definitely perf-sensitive.